### PR TITLE
8247924: Improve javadoc of Foreign Memory Access API (part two)

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch addresses some misc issues in the foreign memory access javadoc, following some feedback from Brian (thanks!).

Things to look out for:

* the package-level javadoc has been tweaked so that we first introduce the entities, then the example - also there's a bit more text to make things clearer
* the javadoc section in MemorySegment on confinement has been moved at the end - since it's fairly advanced
* the toplevel javadoc of `MemoryLayout` has been reworked pretty substantially; it now has a leading example which shows how to go from a C decl to a layout; that same example is then reused throughout the javadoc to explain layout paths

When doing this, I realized that the section on layout paths has several mistakes - from code mistakes to a section which was repeated twice (!!) no idea how that could have gone through all the review rounds (it's not a merge artifact - same issue is in jdk/jdk).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8247924](https://bugs.openjdk.java.net/browse/JDK-8247924): Improve javadoc of Foreign Memory Access API ⚠️ Title mismatch between PR and JBS.

### Reviewers
 * Jorn Vernee ([jvernee](@JornVernee) - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/211/head:pull/211`
`$ git checkout pull/211`

### CODE REVIEW

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java [hunk 1]
```diff
@@ -103,53 +124,30 @@
...
-SequenceLayout seq = MemoryLayout.ofSequence(5,
-    MemoryLayout.ofStruct(
-        MemoryLayout.ofPaddingBits(32),
-        MemoryLayout.ofValueBits(32, ByteOrder.BIG_ENDIAN).withName("value")
-));
- * }</pre></blockquote>
- *
- * We can obtain the offset, in bits, of the member layout named <code>value</code> from <code>seq</code>, as follows:
+ * For instance, given the {@code taggedValues} layout instance constructed as above, we can obtain the offset,
+ * in bits, of the <em>first</em> member layout named <code>value</code>, as follows:
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java:128]: Shouldn't it say 'the second member layout'? ('kind' is the first)

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java [hunk 2]
```diff
@@ -103,53 +124,30 @@
...
+MemoryLayout taggedValuesWithHole = taggedValues.map(l -> MemoryLayout.ofPadding(32),
+                                            PathElement.sequenceElement(), PathElement.groupElement("value"));
  * }</pre></blockquote>
  *
  * That is, the above declaration is identical to the following, more verbose one:
  * <blockquote><pre>{@code
-MemoryLayout newSeq = MemoryLayout.ofSequence(5,
+MemoryLayout taggedValuesWithHole = MemoryLayout.ofSequence(5,
     MemoryLayout.ofStruct(
+        MemoryLayout.ofPaddingBits(8, ByteOrder.NATIVE_ORDER).withName("tag").
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java:150]:
```suggestion
        MemoryLayout.ofPaddingBits(8, ByteOrder.NATIVE_ORDER).withName("kind").
```

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java [hunk 3]
```diff
@@ -25,10 +25,21 @@
...
+ * <p>
+ * The key abstractions introduced by this package are {@link jdk.incubator.foreign.MemorySegment} and {@link jdk.incubator.foreign.MemoryAddress}.
+ * The first models a contiguous memory region, which can reside either inside or outside the Java heap; the latter models an address - which can
+ * sometimes be expressed as an offset into a given segment. A memory address represents the main access coordinate of a memory access var handle, which can be obtained
+ * using the combinator methods defined in the {@link jdk.incubator.foreign.MemoryHandles} class. Finally, the {@link jdk.incubator.foreign.MemoryLayout} class
+ * hierarchy enables description of <em>memory layouts</em> and basic operations such as computing the size in bytes of a given
+ * layout, obtain its alignment requirements, and so on. Memory layouts also provide an alternate, more abstract way, to produce
+ * memory access var handles, e.g. using <a href="MemoryLayout.html#layout-paths"><em>layout paths</em></a>.
+ *
+ * For example, to allocate an off-heap memory region big enough to hold 10 values of the primitive type {@code int}, and fill it values
```
Comment (JornVernee) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:38]:
```suggestion
 * For example, to allocate an off-heap memory region big enough to hold 10 values of the primitive type {@code int}, and fill it with values
```

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java (hunk 1 above)
Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java:128]: first in the sequence

Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryLayout.java:128]: I've tried to rectified this in the latest revision, see if that works better

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java [hunk 4]
```diff
@@ -39,7 +50,9 @@
  * }</pre>
  *
  * Here we create a var handle, namely {@code intHandle}, to manipulate values of the primitive type {@code int}, at
- * a given memory location. We then create a <em>native</em> memory segment, that is, a memory segment backed by
+ * a given memory location. Also, {@code intHandle} is stored in a {@code static} and {@code final} field, to achieve
+ * better performances and allow for inlining of the memory access operation through the {@link java.lang.invoke.VarHandle}
```
Comment (PaulSandoz) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/package-info.java:54]: s/performances/performance

**Pull request closed without merging**